### ISSUE TITLE
gitcommit adaption

### ISSUE THREAD

**Issue by Konfekt** [qadzek/vim-md-link#5]
For what it's worth (maybe put it in a wiki?), here's an adaption to gitcommit messages, moving the links into a link section before any other comments on writing the buffer:

```vim
autocmd vimrcFileTypeGitcommit BufWrite <buffer>
            \ TagLinks |
            \ call mdlink#ConvertAllLinks() |
            \ call PutLinksBeforeComments()

command! -buffer -range=% -bar TagLinks silent keeppatterns <line1>,<line2>substitute#\v%(\[[^\]]+\]\(\s*)@<!\zs<https?://([^.]+)\S+#[\1](\0)#giep

" needs to be :sourced only once per session.
silent! function PutLinksBeforeComments()
    mark `
    normal! gg
    let hash_start = search('^#', 'nW')
    if hash_start == 0 || getline(hash_start) ==? '## Links'
        return
    endif

    let links_start = search('^## Links$', 'nW')
    if links_start == 0
        return
    endif

    let move_start = links_start
    call append(line('$'), '')
    let move_end = line('$')

    keepjumps execute move_start . ',' . move_end . 'move ' . (hash_start - 1)
    normal! g``

    let l:last_non_blank_line = prevnonblank(line('$'))
    if l:last_non_blank_line < line('$')
        execute (l:last_non_blank_line + 1) . ',$delete'
    endif
endfunction
```

**Comment by Konfekt** [comment]
I am sorry, the autocmd group `vimrcFileTypeGitcommit` is a personal leftover, there would be a preceding buffer-local augroup definition like

```vim
augroup vimrcFileTypeGitcommit
  autocmd! * <buffer>
augroup END
```

The idea of Taglinks is to convert all https into markdown links, so that your plug-in can treat them.

One motiviation of using your plug-in is that sometimes one lazily pastes long https into a message, but  these better be put at the end, and some reference inline.

**Comment by Konfekt** [comment]
By the way, the wiki does not yet show up here

**Comment by Konfekt** [comment]
I think it does not hurt to provide sensible stubs to other common text ftplugins. Maybe Taglinks (corrected) can be added generically?  I think this preprocessing is what https://github.com/qadzek/vim-md-link/issues/1 has in mind, and also what sprang to mine when I heard of the existence of your plug-in

**Comment by Konfekt** [comment]
Yes, at the moment it works as described in the README

```md
# Notes

[Vim](https://www.vim.org) and [Neovim](https://neovim.io) are text editors.
```

will be turned into

```md
# Notes

[Vim][0] and [Neovim][1] are text editors.

## Links

[0]: https://www.vim.org
[1]: https://neovim.io
```

But true, in text messages, like mail, gitcommits, ..., more common is a simple bracketed number.

So the taglinks preprocesses  the links for md-link. 

You propose a command to postprocess them, removing the tag?
I have not so strong opinions on it; the domain name can make the text more readable

**Comment by Konfekt** [comment]
Take a look at the latest wiki modifications; it was a bit of work to make it work

**Comment by Konfekt** [comment]
> In both snippets, there is a hardcoded heading value in let links_start = search('^## Links$', 'nW').

Thank you very much, I missed updating the function

**Comment by Konfekt** [comment]
Yes, seems to work:

```markdown
[github.com][1]

[github.com][0]

Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor 
incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis      
nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.   

[github.com][2]

Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu 
fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in  
culpa qui officia deserunt mollit anim id est laborum.                          

Links:

[0]: https://github.com/qadzek/vim-md-link/issues/5
[1]: https://github.com/qadzek/vim-md-link/issues/5
[2]: https://github.com/qadzek/vim-md-link/issues/5

-- 
Signature

```

using

```vim
let b:md_link_heading = 'Links:'
let b:md_link_heading_before = '^-- $'
let s:body_range = '1/^\s*$/,/' . b:md_link_heading_before.'\|^\V'.escape(b:md_link_heading,'\').'\m' . '$/-1'

command! -buffer -range=% -bar TagLinks silent keepjumps keeppatterns <line1>,<line2>substitute#\v%(\[[^\]]+\]\(\s*)@<!\zs<https?://(%([a-zA-Z0-9-]+\.)+[a-zA-Z]{2,})\S+#[\1](\0)#giep
autocmd BufWrite <buffer>
            \ exe s:body_range . 'TagLinks'|
            \ exe s:body_range . 'MdLinkConvertRange'
```

**Comment by Konfekt** [comment]
For gitcommit less so:

```


[github.com][0]


# Immetti il messaggio di commit per le modifiche. Le righe che iniziano
# con '#' saranno ignorate e un messaggio vuoto interromperà il commit.
#
# Sul branch master
# Il tuo branch è aggiornato rispetto a 'origin/master'.
#
# Modifiche di cui verrà eseguito il commit:
#	modificato:             .vim/ftplugin/gitcommit_md-link.vim
#
# ------------------------ >8 ------------------------

Links:

[0]: https://github.com/qadzek/vim-md-link/issues/5
```

using

```vim
let b:md_link_heading = 'Links:'
let b:md_link_heading_before = '^#'
let s:body_range = '1,/' . b:md_link_heading_before.'\|\V\^'.escape(b:md_link_heading,'\') . '\$/-1'

command! -buffer -range=% -bar TagLinks silent keepjumps keeppatterns <line1>,<line2>substitute#\v%(\[[^\]]+\]\(\s*)@<!\zs<https?://(%([a-zA-Z0-9-]+\.)+[a-zA-Z]{2,})\S+#[\1](\0)#giep
command! -buffer -range=% LinkConvertRange <line1>,<line2>TagLinks | <line1>,<line2>MdLinkConvertRange
autocmd BufWrite <buffer> exe s:body_range..'LinkConvertRange'
```

**Comment by Konfekt** [comment]
Oh, that's good to know. Yes, that works.

**Comment by Konfekt** [comment]
That's great. Would you mind allowing a range to MdLinkProcessConvert? For example, to spare links in the mail header. Also in Git commit only up to the first comment. So the `s:body_range` still has its use.

Also, how about allowing concatenating commands by adding `-bar` to their definition ?

**Comment by Konfekt** [comment]
> Should this be added to every command? Thanks for the suggestion, I wasn't really aware of -bar.

Yes, indeed

>  I am not sure how to pass a range from one function to another?

Does

```vim
function! FirstFunction(...) range
  " Capture the range into variables
  let l:startline = a:firstline
  let l:endline = a:lastline

  " Do something with the range, if needed
  " ...

  " Call the second function with the range
  execute l:startline.','.l:endline.'call SecondFunction()'
endfunction

function! SecondFunction() range
  " This function now has access to the range
  " Do something with the range
  for i in range(a:firstline, a:lastline)
    " Process each line
    " ...
  endfor
endfunction
```

help?

**Comment by Konfekt** [comment]
Yes, it works flawlessly, thank you very much.
I updated the wiki accordingly.

Maybe in the future, comments could be skipped, say by an exclusion regex `^>`. 

For example, in a mail, one rarely wants to reference all the URLs in the cited text.
One can restrict the range to the first citation, but maybe the text continues afterward:

```

[0]

John Doe said:
>
> [1]

Links:

[0]: https://github.com/qadzek/vim-md-link/tree/dev
[1]: https://github.com/qadzek/vim-md-link/tree/main
```

**Comment by Konfekt** [comment]
> I am not sure if let s:body_range still needed, in both extensions?

Yes, for the mails only the links in the body need conversion, not in the subject header say. 
For commits, not those in comments.

> And for ft=gitcommit, is PutLinksBeforeComments() still required? let b:md_link_heading_before should take care of that part.

As you explained above, `b:md_link_heading_before` looks from below, but the links are placed before the comments

**Comment by Konfekt** [comment]
Great! I'll use this starting from now.

> b:md_link_skip_line

You could suggest, or even default to, `&commentstring`

**Comment by Konfekt** [comment]
I am trying

```vim
let b:md_link_skip_line = substitute('\V\^'..escape(&commentstring, '\')..'\$', '%s', '\.\*', 'g')
```

**Comment by Konfekt** [comment]
I adapted the Wiki accordingly; the setup lines are becoming [fewer](https://github.com/qadzek/vim-md-link/wiki/Git-Commit-Message)

**Comment by Konfekt** [comment]
> This has now been implemented: https://github.com/qadzek/vim-md-link/commit/1764dbc7d7f9842719574340c6642b5994797722

While this is a good default, it is not perfect, as the commentstring uses spacing after the comment sign in, say shell or mail  (and likely many other) files, and thus does not recognize, say `##` or `>>`. 
Since most likely no comment signs depend on white space, maybe just strip the space before `%s` to make it more robust.

**Comment by Konfekt** [comment]
Yes, that will work. But  `///` will not work for Javascript, though still a comment. Well, it's a finesse, but I doesn't cost much calling `substitute(b:md_link_skip_line, '\s', '', g)`

**Comment by Konfekt** [comment]
Surely, but what do I have to try, and what to change? I am mainly using mdlink as documented in the wiki

**Comment by Konfekt** [comment]
Maybe you could indicate how the wiki needs to be updated?

**Comment by Konfekt** [comment]
> MdLinkProcessConvert has been removed, so it should be replaced in your snippets, both in the if exists and in the autocmd. You are converting a range (s:body_range), so you should be able to replace it by MdLinkConvertRange.

Okay, so `MdLinkProcessConvert` is now `MdLinkConvertRange`? Feel free to adapt the Wiki

**Comment by Konfekt** [comment]
Okay, thank you. This is on the dev branch that I will switch to using your adaptions of the wiki.

**Comment by Konfekt** [comment]
Thank you for the heads-up. All the changes seem sensible. I hope to have caught all affected places in my local config

**Comment by Konfekt** [comment]
Regarding the autocmd, it looks fine, but somehow fires later than defining it in a ftplugin file.
Therefore the previous check for the existence of `:Link` command became for the default file types

```vim
if !exists('g:loaded_link') | finish | endif
if exists('g:link_enabled_filetypes') && index(g:link_enabled_filetypes, &l:filetype) == -1 | finish | endif
```

respectively for the others

```vim
if !exists('g:link_enabled_filetypes') || index(g:link_enabled_filetypes, &l:filetype) == -1 | finish | endif
```

**Comment by Konfekt** [comment]
> g:link_enabled_filetypes not existing is allowed. Currently, if the user doesn't set g:link_enabled_filetypes, the plugin will be activated for the default filetypes: let s:default_enabled_filetypes = [ 'markdown', 'vimwiki', 'mail', 'text' ].

Exactly, that's why for these filetypes additional checks are needed, as implemented above.

> did notice that e.g. when opening [foo.py](http://foo.py/) and executing :set ft=markdown, the plugin isn't working, while I expected the autocmd would fire. Not sure if that's related.

That must be a separate issue; for me on Ubuntu 22.04 with 9.1.369 it fires when switching, say from TomL files.

> I am not really sure how to implement your two snippets.

These are maybe useful for the snippets in the wiki, but now seem overkill
